File: ExMCL Forge Mods/src/main/java/com/n9mtq4/exmcl/forgemods/utils/Downloader.java
package com.n9mtq4.exmcl.forgemods.utils;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.nio.channels.Channels;
import java.nio.channels.ReadableByteChannel;

/**
 * Created by will on 2/14/16 at 11:06 PM.
 *
 * @author Will "n9Mtq4" Bresnahan
 */
public class Downloader {
	
	/**
	 * Downloads a url to a file
	 * 
	 * TODO: this code is a MESS
	 * 
	 * @param url the url
	 * @param file the file
	 * */
	public static void downloadFile(String url, File file) {
		
		URL website = null;
		try {
			website = new URL(url);
		} catch (MalformedURLException e) {
			e.printStackTrace();
		}
		ReadableByteChannel rbc = null;
		try {
			if (website != null) {
				rbc = Channels.newChannel(website.openStream());
			}
		} catch (IOException e1) {
			e1.printStackTrace();
		}
		FileOutputStream fos = null;
		try {
			fos = new FileOutputStream(file);
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}
		try {
			if (fos != null) {
				fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);
			}
			if (fos != null) {
				fos.close();
			}
			if (rbc != null) {
				rbc.close();
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
		
	}
	
}

File: ExMCL Forge Mods/src/main/kotlin/com/n9mtq4/exmcl/forgemods/utils/Downloader.kt
/*
 * MIT License
 *
 * Copyright (c) 2016 Will (n9Mtq4) Bresnahan
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.n9mtq4.exmcl.forgemods.utils

import com.n9mtq4.kotlin.extlib.pstAndGiven
import java.io.File
import java.io.FileOutputStream
import java.net.URL
import java.nio.channels.Channels

/**
 * Created by will on 3/10/16 at 8:47 PM.
 *
 * @author Will "n9Mtq4" Bresnahan
 */

/**
 * Downloads a url to a file
 * 
 * @param url the url
 * @param file the file
 * @return if the download was successful
 * */
internal fun downloadFile(url: String, file: File) = pstAndGiven(false) {
	
	val website = URL(url)
	val rbc = Channels.newChannel(website.openStream())
	val fos = FileOutputStream(file)
	
	fos.channel.transferFrom(rbc, 0, Long.MAX_VALUE)
	fos.close()
	rbc.close()
	
	true
	
}
